src/EditorFeatures/CSharp/Formatting/CSharpFormattingInteractionService.cs

```diff
@@ -89,56 +89,54 @@ public bool SupportsFormattingOnTypedCharacter(Document document, char ch)
+            var formattingSpan = CommonFormattingHelpers.GetFormattingSpan(parsedDocument.Root, span);
 
-            var services = document.Project.Solution.Workspace.Services;
-            return Formatter.GetFormattedTextChanges(root, SpecializedCollections.SingletonEnumerable(formattingSpan), services, options, cancellationToken).ToImmutableArray();
+            return Task.FromResult(Formatter.GetFormattedTextChanges(parsedDocument.Root, SpecializedCollections.SingletonEnumerable(formattingSpan), parsedDocument.LanguageServices.WorkspaceServices, options, cancellationToken).ToImmutableArray());
         }
 
-        public async Task<ImmutableArray<TextChange>> GetFormattingChangesOnPasteAsync(Document document, ITextBuffer textBuffer, TextSpan textSpan, CancellationToken cancellationToken)
+        public Task<ImmutableArray<TextChange>> GetFormattingChangesOnPasteAsync(Document document, ITextBuffer textBuffer, TextSpan textSpan, CancellationToken cancellationToken)
         {
+            var parsedDocument = ParsedDocument.CreateSynchronously(document, cancellationToken);
             var fallbackOptions = _globalOptions.GetCSharpSyntaxFormattingOptions();
-            var options = _indentationManager.GetInferredFormattingOptions(textBuffer, _editorOptionsFactory, document.Project.LanguageServices, fallbackOptions, explicitFormat: true);
-            var service = document.GetRequiredLanguageService<ISyntaxFormattingService>();
-            var documentSyntax = await ParsedDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
-            return service.GetFormattingChangesOnPaste(documentSyntax, textSpan, options, cancellationToken);
+            var options = _indentationManager.GetInferredFormattingOptions(textBuffer, _editorOptionsFactory, parsedDocument.LanguageServices, fallbackOptions, explicitFormat: true);
+            var service = parsedDocument.LanguageServices.GetRequiredService<ISyntaxFormattingService>();
+            return Task.FromResult(service.GetFormattingChangesOnPaste(parsedDocument, textSpan, options, cancellationToken));
         }
 
-        Task<ImmutableArray<TextChange>> IFormattingInteractionService.GetFormattingChangesOnReturnAsync(
-            Document document, int caretPosition, CancellationToken cancellationToken)
+        public Task<ImmutableArray<TextChange>> GetFormattingChangesOnReturnAsync(Document document, int caretPosition, CancellationToken cancellationToken)
             => SpecializedTasks.EmptyImmutableArray<TextChange>();
 
-        public async Task<ImmutableArray<TextChange>> GetFormattingChangesAsync(Document document, ITextBuffer textBuffer, char typedChar, int position, CancellationToken cancellationToken)
+        public Task<ImmutableArray<TextChange>> GetFormattingChangesAsync(Document document, ITextBuffer textBuffer, char typedChar, int position, CancellationToken cancellationToken)
```

Comment:
sam has an attribute about performance.  we should mark these methods as ones we want to keep running synchronously.

alternatively we could have an ISynchronousFormattingInteractionService and just have our partners use the async version until they can port to the synchronous one.

Reply:
This is just for TypeScript. I tried adding another interface but it leads to duplication. Better just fix TypeScript and then make this synchronous. Filed issue: https://github.com/dotnet/roslyn/issues/62003